File: LatinReader/core/src/test/java/com/ericmschmidt/latinreader/datamodel/BookTest.java
package com.ericmschmidt.latinreader.datamodel;

import org.junit.Before;
import org.junit.Test;

import static com.google.common.truth.Truth.assertThat;

import com.ericmschmidt.classicsreader.datamodel.Book;

public class BookTest {

    Book testBook;
    String expectedLine1 = "This is line 1";
    String expectedLine2 = "This is line 2";
    String expectedLine3 = "This is line 3";

    @Before
    public void setUp() {
        testBook = new Book(1234);
        testBook.addLines(expectedLine1);
        testBook.addLines(expectedLine2);
        testBook.addLines(expectedLine3);
    }


    @Test
    public void testGetLine() {

        // Verify that it gets the correct line.
        String actualResult1 = testBook.getLine(0);
        assertThat(actualResult1).contains(expectedLine1);

        // Verify that passing in a negative number returns the first line.
        String actualResult2 = testBook.getLine(-1);
        assertThat(actualResult2).contains(expectedLine1);

        // Verify that passing in a number larger than the number
        // of lines returns null.
        String actualResult3 = testBook.getLine(5);
        assertThat(actualResult3).isNull();
    }

    @Test
    public void testGetLines() {

        // Verify that it gets the correct lines.
        String actualResult1 = testBook.getLines(0, 2);
        assertThat(actualResult1).contains(expectedLine1);
        assertThat(actualResult1).contains(expectedLine2);

        // Verify that passing in a negative number as the position returns
        // lines from the beginning of the book.
        String actualResult2 = testBook.getLines(-2, 2);
        assertThat(actualResult2).contains(expectedLine1);
        assertThat(actualResult2).contains(expectedLine2);

        // Verify that passing in a number greater than the total lines
        // returns the remaining lines in the book.
        String actualResult3 = testBook.getLines(1, 10);
        assertThat(actualResult3).contains(expectedLine2);
        assertThat(actualResult3).contains(expectedLine3);
    }

}

File: LatinReader/core/src/test/java/com/ericmschmidt/classicsreader/data/BookTest.kt
package com.ericmschmidt.classicsreader.data

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Before
import org.junit.Test

class BookTest {

    private lateinit var book: Book

    @Before
    fun setUp() {
        book = Book(1)
        book.addLines("line 1")
        book.addLines("line 2")
        book.addLines("line 3")
    }

    @Test
    fun addLines() {
        val initialLineCount = book.getLineCount()
        book.addLines("line 4")
        assertEquals(initialLineCount + 1, book.getLineCount())
    }

    @Test
    fun getLine() {
        assertEquals("line 1", book.getLine(0))
        assertEquals("line 2", book.getLine(1))
        assertEquals("line 3", book.getLine(2))
    }

    @Test
    fun getLine_invalidPosition_returnsNull() {
        assertNull(book.getLine(3))
        assertNull(book.getLine(10))
    }

    @Test
    fun getLine_negativePosition_returnsFirstLine() {
        assertEquals("line 1", book.getLine(-1))
    }

    @Test
    fun getLines() {
        assertEquals("line 1\nline 2\n", book.getLines(0, 2))
        assertEquals("line 2\nline 3\n", book.getLines(1, 2))
    }

    @Test
    fun getLines_offsetExceedsSize_returnsUpToEnd() {
        assertEquals("line 1\nline 2\nline 3\n", book.getLines(0, 10))
    }

    @Test
    fun getLines_negativePosition_startsFromBeginning() {
        assertEquals("line 1\nline 2\n", book.getLines(-1, 2))
    }

    @Test
    fun getLines_empty() {
        val emptyBook = Book(2)
        assertEquals("", emptyBook.getLines(0, 1))
    }

    @Test
    fun getLineCount() {
        assertEquals(3, book.getLineCount())
    }
}